pkg/controller/common/rollout/workloads/deployment_controller.go

```diff
@@ -73,15 +73,19 @@ func (c *DeploymentController) VerifySpec(ctx context.Context) (bool, error) {
 	defer func() {
 		if verifyErr != nil {
 			klog.Error(verifyErr)
-			c.recorder.Event(c.parentController, event.Warning("VerifyFailed", verifyErr))
+			if c.recorder != nil && c.parentController != nil {
+				c.recorder.Event(c.parentController, event.Warning("VerifyFailed", verifyErr))
+			}
 		}
 	}()
 
 	// check if the rollout spec is compatible with the current state
 	targetTotalReplicas, verifyErr := c.calculateTargetTotalSize(ctx)
 	if verifyErr != nil {
 		// do not fail the rollout just because we can't get the resource
-		c.rolloutStatus.RolloutRetry(verifyErr.Error())
+		if c.rolloutSpec != nil {
```

Comment:
when will the spec be nil?

Reply:
Same reason as above. Should I remove the check nil logic?

Reply:
yeah, we should not check (and allow) conditions that should not happen.

Reply:
Okay, removed.